src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
-#include "cFE_util.h"
 #include "osfilesys.h"
+#include "tar.h"
 
-struct FS fs[MAX_NUM_FS];
+struct fs filesystems[MAX_NUM_FS];
+struct fs *openfs = &filesystems[0];;
 struct fsobj files[MAX_NUM_FILES];
+struct cos_compinfo *ci;
+
+struct fd fd_tbl[OS_MAX_NUM_OPEN_FILES + 1];
+char copy_buffer[F_PART_DATA_SIZE];
 
 /*
  * Notes on this version:
- * This is not the final implementation of a file system for the cFE on COS
- * For now the file system is read only and staticly allocates for fsobj and FS
- * The next step is to write a dumb allocator/deallocator and use it for
- * new fs, fsobj, and file data
- * I have been intentional to restrict memory specific details to this file
- * osfilesys.h and osfiles.c 'should' not care about how FS, fsobj, and data
- * are stored in memory.
+ * It provides all of the functionality required by the cFE and more.  It does
+ * not fully satisfy all of the requirements for posix in general. It passes all
+ * of cFE's unit tests, and all of the testing and debugging I have thrown at it
+ *
+ * Beware when using struct dirent and struct stat.  As a temporary measure,
+ * they are currently generated to comply with glibc rather than musl.
+ *
+ * Future plans:
+ *  - Compile cFE with musl and fix stat and dirent structures
+ *  - modularize into a bare bones filesystem library
+ *  - Refactor rest of code into posix compatible filesystem built on bare bones fs
+ *
  */
 
 /******************************************************************************
-** Tar Level Methods
+** fsobj Level Methods
 ******************************************************************************/
 
-/*
- * Loads the position in memory of linked tar file system
- * Checks for badly linked or no linked tar file.  The names given by
- * the linker are non-intuative so a description of error checking is given:
- * First checks to make sure that symbols have been overwritten by linking
- * process.  Next checks that the size is greater than 0.  Finally checks that
- * the end of the tar is after the start
- */
-uint32 tar_load()
+int32 chk_fd(int32 FD)
 {
-    if (!_binary_cFEfs_tar_start)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    if (! &_binary_cFEfs_tar_size)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    if (&_binary_cFEfs_tar_end < &_binary_cFEfs_tar_start)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-    size    = (size_t) &_binary_cFEfs_tar_size;
-    start   = &_binary_cFEfs_tar_start;
-    end     = &_binary_cFEfs_tar_end;
-    return  OS_FS_SUCCESS;
+    if (FD > OS_MAX_NUM_OPEN_FILES) return OS_FS_ERR_INVALID_FD;
+    if (FD <= 0)                        return OS_FS_ERR_INVALID_FD;
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (!filedes->file) return OS_FS_ERR_INVALID_FD;
+    if (filedes->ino != filedes->file->ino) return OS_FS_ERR_INVALID_FD;
```

Comment:
it would be a legal state of the filesystem if one of both of them were 0 and the other a valid other value.  The real invalid state would be if file->ino and filedes->ino were nonezero and different.  I now explicitly test for both of these and do the right thing in each case.